staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go

```diff
@@ -0,0 +1,276 @@
+		testDeviceName     = "test-device"
+		testResourceSlice1 = "test-resource-slice-1"
+		testResourceSlice2 = "test-resource-slice-2"
+		testResourceSlice3 = "test-resource-slice-3"
+	)
+
+	testCases := []struct {
+		name                   string
+		initialObjects         []runtime.Object
+		inputPoolName          string
+		inputDriverResources   *DriverResources
+		expectedResourceSlices []resourceapi.ResourceSlice
+		expectedError          error
+	}{
+		{
+			name: "single ResourceSlice Existing, so we just update it",
+			initialObjects: []runtime.Object{
+				newResourceSlice(testResourceSlice1, testNode, testDriveName, testPoolName),
+			},
+			inputPoolName: testPoolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					testPoolName: {
+						Devices: []resourceapi.Device{
+							{
+								Name: testDeviceName,
+							},
+						},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*newExpectResourceSlice(testResourceSlice1, testNode, string(testNodeUID), testDriveName, testPoolName, testDeviceName, false),
+			},
+		},
+		{
+			name: "multiple ResourceSlices Existing, so we update and delete",
+			initialObjects: []runtime.Object{
+				newResourceSlice(testResourceSlice1, testNode, testDriveName, testPoolName),
+				newResourceSlice(testResourceSlice2, testNode, testDriveName, testPoolName),
+				newResourceSlice(testResourceSlice3, testNode, testDriveName, testPoolName),
+			},
+			inputPoolName: testPoolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					testPoolName: {
+						Devices: []resourceapi.Device{
+							{
+								Name: testDeviceName,
+							},
+						},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*newExpectResourceSlice(testResourceSlice1, testNode, string(testNodeUID), testDriveName, testPoolName, testDeviceName, false),
+			},
+		},
+		{
+			name:           "no existing ResourceSlices, so we create",
+			initialObjects: []runtime.Object{},
+			inputPoolName:  testPoolName,
+			inputDriverResources: &DriverResources{
+				Pools: map[string]Pool{
+					testPoolName: {
+						Devices: []resourceapi.Device{
+							{
+								Name: testDeviceName,
+							},
+						},
+					},
+				},
+			},
+			expectedResourceSlices: []resourceapi.ResourceSlice{
+				*newExpectResourceSlice(testNode+"-"+testDriveName+"-", testNode, string(testNodeUID), testDriveName, testPoolName, testDeviceName, true),
+			},
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			_, ctx := ktesting.NewTestContext(t)
+			// Create kubeClient
+			kubeClient := fake.NewSimpleClientset(test.initialObjects...)
+
+			// Create controller
+			ctrl, informerFactory := setupController(ctx, testNode, testNodeUID, testDriveName,
+				kubeClient, test.inputDriverResources)
+
+			// Start informer factory
+			informerFactory.Start(ctx.Done())
+			informerFactory.WaitForCacheSync(ctx.Done())
```

Comment:
How about moving the initialization logic for the informer factory into the `setupController()` function? That way, we don't need to return the `informerFactory` directly from `setupController()` function.

```suggestion
```

Reply:
I set the sorting here because the objects obtained from the informer may not be the same every time, which will cause unstable test cases. And I understand that when only the first one is obtained and the subsequent ones are deleted, we should ensure the order

Reply:
thanks!